pkg/gcp/ocpgwdeployer.go

```diff
@@ -0,0 +1,285 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package gcp
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/cloud-prepare/pkg/ocp"
+	"google.golang.org/api/compute/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+)
+
+type ocpGatewayDeployer struct {
+	gcp          *gcpCloud
+	msDeployer   ocp.MachineSetDeployer
+	instanceType string
+	image        string
+}
+
+// NewOcpGatewayDeployer returns a GatewayDeployer capable deploying gateways using OCP
+// If the supplied cloud is not a gcpCloud, an error is returned
+func NewOcpGatewayDeployer(cloud api.Cloud, msDeployer ocp.MachineSetDeployer, instanceType, image string) (api.GatewayDeployer, error) {
+	gcp, ok := cloud.(*gcpCloud)
+	if !ok {
+		return nil, errors.New("the cloud must be GCP")
+	}
+
+	return &ocpGatewayDeployer{
+		gcp:          gcp,
+		msDeployer:   msDeployer,
+		instanceType: instanceType,
+		image:        image,
+	}, nil
+}
+
+func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, reporter api.Reporter) error {
+	var currentGWInstanceList []*compute.Instance
+
+	reporter.Started(messageCreateExtFWRules)
+
+	externalIngress := newExternalFirewallRules(d.gcp.projectID, d.gcp.infraID, input.PublicPorts)
+	if err := d.gcp.openPorts(externalIngress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened External ports %q with firewall rule %q on GCP",
+		formatPorts(input.PublicPorts), externalIngress.Name)
+
+	reporter.Started(messageRetrieveZones)
+
+	zones, err := d.gcp.client.ListZones()
+	if err != nil {
+		reporter.Failed(err)
+		return fmt.Errorf("failed to list the zones in the project %q. %v", d.gcp.projectID, err)
```

Comment:
You augment the return error with additional context - perhaps we should do the same with the `reporter`? Applies to other places as well

```suggestion
		return reportFailure(reporter, err, "failed to list the zones in the project %q", d.gcp.projectID)
		
	...
	func reportFailure(reporter api.Reporter, failure error, format string, args ...string) {
	    err  := errors.WithMessagef(failure, format, args)
	    reporter.Failed(err)
	    
	    return err
	}
```